src/providers/tpm_provider/mod.rs

```diff
@@ -0,0 +1,461 @@
+
+        let error_closure = |e| {
+            error!("Error storing a mapping: {}.", e);
+            Err(ResponseStatus::KeyIDManagerError)
+        };
+        if store_handle
+            .remove(&key_triple)
+            .or_else(error_closure)?
+            .is_none()
+        {
+            error!(
+                "Key triple \"{}\" does not exist in the Key ID Manager.",
+                key_triple
+            );
+            Err(ResponseStatus::KeyDoesNotExist)
+        } else {
+            Ok(ResultDestroyKey {})
+        }
+    }
+
+    fn asym_sign(&self, app_name: ApplicationName, op: OpAsymSign) -> Result<ResultAsymSign> {
+        let key_name = op.key_name;
+        let hash = op.hash;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+
+        let store_handle = self.key_id_store.read().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let len = hash.len();
+        if len > 64 {
+            error!("The buffer given to sign is too big. Its length is {} and maximum authorised is 64.", len);
+            return Err(ResponseStatus::PsaErrorInvalidArgument);
+        }
+
+        let password_context = get_password_context(&*store_handle, key_triple)?;
+
+        let signature = esapi_context
+            .sign(
+                password_context.context,
+                &password_context.auth_value,
+                &hash,
+            )
+            .or_else(|e| {
+                error!("Error signing: {}.", e);
+                Err(ResponseStatus::PsaErrorHardwareFailure)
+            })?;
+
+        Ok(ResultAsymSign {
+            signature: signature.signature,
+        })
+    }
+
+    fn asym_verify(&self, app_name: ApplicationName, op: OpAsymVerify) -> Result<ResultAsymVerify> {
+        let key_name = op.key_name;
+        let hash = op.hash;
+        let signature = op.signature;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+
+        let store_handle = self.key_id_store.read().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let len = hash.len();
+        if len > 64 {
+            error!("The buffer given to sign is too big. Its length is {} and maximum authorised is 64.", len);
+            return Err(ResponseStatus::PsaErrorInvalidArgument);
+        }
+
+        let signature = Signature {
+            scheme: AsymSchemeUnion::RSASSA(TPM2_ALG_SHA256),
+            signature,
+        };
+
+        let password_context = get_password_context(&*store_handle, key_triple)?;
+
+        esapi_context
+            .verify_signature(password_context.context, &hash, signature)
+            .or_else(|e| {
```

Comment:
I suggest doing a mapping between `Tss2ResponseCode` or the kind and the Parsec `ResponseStatus` - an `impl From<Tss2......` would let you just do `?` - but maybe have this as another issue

Reply:
Yes that is a good idea and something we should also do for the PKCS 11 provider. Currently both provider just return the generic "hardware failure" error. We will need that for the 1.0 milestone. Will raise an issue.

Reply:
This is actually under existing issue #72